core/src/com/unciv/logic/map/UnitMovementAlgorithms.kt

```diff
@@ -213,47 +228,50 @@ class UnitMovementAlgorithms(val unit: MapUnit) {
+                        }
+                        path.reverse() // and reverse in order to get the list in chronological order
+                        pathfindingCache.setShortestPathCache(destination, path)
+                        return path
                     } else {
                         if (movementTreeParents.containsKey(reachableTile)) continue // We cannot be faster than anything existing...
                         if (!isUnknownTileWeShouldAssumeToBePassable(reachableTile) &&
-                                 !canMoveTo(reachableTile)) continue // This is a tile that we can't actually enter - either an intermediary tile containing our unit, or an enemy unit/city
+                                !canMoveTo(reachableTile)) continue // This is a tile that we can't actually enter - either an intermediary tile containing our unit, or an enemy unit/city
                         movementTreeParents[reachableTile] = tileToCheck
                         newTilesToCheck.add(reachableTile)
                     }
                 }
             }
 
-            if (distanceToDestination.isNotEmpty()) {
-                val path = mutableListOf(destination) // Traverse the tree upwards to get the list of tiles leading to the destination,
-                // Get the tile from which the distance to the final tile in least -
-                // this is so that when we finally get there, we'll have as many movement points as possible
-                var intermediateTile = distanceToDestination.minByOrNull { it.value }!!.key
-                while (intermediateTile != currentTile) {
-                    path.add(intermediateTile)
-                    intermediateTile = movementTreeParents[intermediateTile]!!
-                }
-                path.reverse() // and reverse in order to get the list in chronological order
-                return path
+            if (newTilesToCheck.isEmpty()) {
+                // there is NO PATH (eg blocked by enemy units)
+                pathfindingCache.setShortestPathCache(destination, emptyList())
+                return emptyList()
             }
 
-            if (newTilesToCheck.isEmpty()) return emptyList() // there is NO PATH (eg blocked by enemy units)
-
             // add newTilesToCheck to visitedTiles so we do not path over these tiles in a later iteration
             visitedTiles.addAll(newTilesToCheck)
             // no need to check tiles that are surrounded by reachable tiles, only need to check the edgemost tiles.
             // Because anything we can reach from intermediate tiles, can be more easily reached by the edgemost tiles,
             // since we'll have to pass through an edgemost tile in order to reach the destination anyway
-            tilesToCheck = newTilesToCheck.filterNot { tile -> tile.neighbors.all { it in newTilesToCheck || it in tilesToCheck } }
+            tilesToCheck = newTilesToCheck.filterNot { tile -> tile.neighbors.all { it in visitedTiles } }
```

Comment:
`visitedTiles` is a `HashSet` that contains the same values as both `newTilesToCheck` and `tilesToCheck` so this check can be performed faster.